### PR TITLE
Documentation on using DID with single Page applications

### PR DESCRIPTION
**Pull request by CrowdHailer**

### CODE REVIEW

File: docs/spa/index.md [hunk 1]
```diff
@@ -0,0 +1,183 @@
+---
+layout: doc
+title: SPA integration guide
+abstract: How to integrate Single Page Webapps with DID.app
+---
+
```
Comment (richardesigns) [inline docs/spa/index.md:6]: Please can we have another line break in here between the first sentence and the abstract?  Presume just add a line in MD>

File: docs/spa/index.md [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+Single Page WebApps (SPAs) can use DID.app can authenticate users via OpenID Connect.
```
Comment (richardesigns) [inline docs/spa/index.md:7]:
```suggestion
Single Page WebApps (SPAs) can use DID.app for authentication using OpenID Connect.
```

File: docs/spa/index.md [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+Authenticating End Users takes two steps.
```
Comment (richardesigns) [inline docs/spa/index.md:8]:
```suggestion
There are two steps to authenticating End Users:
```

File: docs/spa/index.md [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+1. Redirect a user to DID.app with an authentication request from your application.
+2. Fetch the user information from the authentication response.
```
Comment (richardesigns) [inline docs/spa/index.md:11]:
```suggestion
2. Fetch the user's information from the authentication response.
```

File: docs/spa/index.md [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+
+Fetching user information for authenticated users can be achieved via two methods.
```
Comment (richardesigns) [inline docs/spa/index.md:13]:
```suggestion
Fetching user information from the authentication response can be done in one of two ways:
```

File: docs/spa/index.md [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+1. Have an API endpoint on your application backend fetch an id token.
```
Comment (richardesigns) [inline docs/spa/index.md:15]:
```suggestion
1. Have an API endpoint on your application's backend to fetch an id token.  Or;
```

File: docs/spa/index.md [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+2. Use PKCE and fetch the id token from within the web app.
```
Comment (richardesigns) [inline docs/spa/index.md:16]:
```suggestion
2. Use PKCE (Proof Key for Code Exchange) and fetch the id token from within the web app.
```

File: docs/spa/index.md [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+
+Use option 1. if you are using cookie based sessions.
```
Comment (richardesigns) [inline docs/spa/index.md:18]:
```suggestion
If you are using cookie based sessions, use option 1.  
```

File: docs/spa/index.md [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+If using DID.app to send invite emails you will have to use option 1.
```
Comment (richardesigns) [inline docs/spa/index.md:19]:
```suggestion
If you are using DID.app to send invite emails you must use option 1.
```

File: docs/spa/index.md [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+
+### Create Authentication Request
+
+Redirect the End User to the authorization url: [https://auth.did.app/oidc/authorize](https://auth.did.app/oidc/authorize).
+Provide the details of the authentication request as query parameters.
+
```
Comment (richardesigns) [inline docs/spa/index.md:25]: Presumably an SPA developer would know what to do here?  How do I send the end user to the URL with the params described?

File: docs/spa/index.md [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+```js
+let fragment = window.location.hash.substring(1)
+let params = new URLSearchParams(fragment)
+let code = params.get("code")
+let error = params.get("error")
+
+window.location.replace("#")
+```
+
+The final line is not required, but can make for a nicer user experience.
```
Comment (richardesigns) [inline docs/spa/index.md:61]: if what?  if there is an error?

File: docs/spa/index.md [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+Replacing the location will remove the code from the browsers navigation bar.
```
Comment (richardesigns) [inline docs/spa/index.md:62]:
```suggestion
Replacing the location will remove the code from the browser's navigation bar.
```

File: docs/spa/index.md [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+
+### Fetch User Data
+
+#### Option 1. Using backend endpoint
```
Comment (richardesigns) [inline docs/spa/index.md:66]:
```suggestion
#### Option 1, using backend API endpoint
```

File: docs/spa/index.md [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+const fetch = require("node-fetch");
+
+let response = await fetch("https://auth.did.app/oidc/token", {
+  method: "POST",
+  body: `client_id=[CLIENT_ID]&client_secret=[CLIENT_SECRET]&code=${code}`
+});
+let data = response.json();
+```
+
+The data will contain your users information.
```
Comment (richardesigns) [inline docs/spa/index.md:82]:
```suggestion
The data will contain the user's information.
```

File: docs/spa/index.md [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+
+```js
+let response = await fetch("https://auth.did.app/oidc/token", {
+  method: "POST",
+  body: `client_id=[CLIENT_ID]&code_verifier=${code_verifier}&code=${code}`
+});
+let data = response.json();
+```
+
+The data will contain your users information.
```
Comment (richardesigns) [inline docs/spa/index.md:112]:
```suggestion
The data will contain the user's information.
```

File: docs/spa/index.md [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+  },
+  "id_token": "COMPACT JWT",
+  "access_token": "COMPACT JWT"
+}
+```
+
+The `sub` is the identifier of a user within your application.
+It is intended to be used as the primary identifier of your user and does not change.
+
+You should not rely on a users email being constant over time.
```
Comment (richardesigns) [inline docs/spa/index.md:133]:
```suggestion
You should not rely on a user's email being constant over time.
```

Comment (richardesigns) [reply docs/spa/index.md:133]: This feels like something that is relevant in all integrations or not?

File: docs/spa/index.md [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+
+At this point authentication is complete.
+It is up to your service to handle things from here.
+
+For backend applications this will involve starting a session for the authenticated user.
+How this is done will depend on the platform, or framework that you are using.
+
+<a name="generating-pkce"></a>
+### Generating code challenge and verifier
+
```
Comment (richardesigns) [inline docs/spa/index.md:143]: What is this section trying to achieve?  I felt like I hadn't understood the relevance of this bit.

File: docs/spa/index.md (hunk 1 above)
Comment (CrowdHailer) [reply docs/spa/index.md:6]: That won't make any difference to the output

File: docs/spa/index.md (hunk 7 above)
Comment (CrowdHailer) [reply docs/spa/index.md:16]: Have I not defined this earlier? If not it should be "Word Word (WW)" at first usage

File: docs/spa/index.md (hunk 10 above)
Comment (CrowdHailer) [reply docs/spa/index.md:25]: Yeah I think this is too framework specific to have an example here. This is where we need to start having Vue.js or other front end guides.

File: docs/spa/index.md (hunk 11 above)
Comment (CrowdHailer) [reply docs/spa/index.md:61]: It just cleans up the URL bar,

File: docs/spa/index.md (hunk 17 above)
Comment (CrowdHailer) [reply docs/spa/index.md:143]: This is basically an appendix to where the function `randomUrlBase64` is used earlier in the post

File: docs/spa/index.md (hunk 1 above)
Comment (richardesigns) [reply docs/spa/index.md:6]: So is this not possible in the constraints of MD?

File: docs/spa/index.md (hunk 10 above)
Comment (richardesigns) [reply docs/spa/index.md:25]: Ah right ok. This wasn't clear is my feedback but now I know it's an appendix then that's quite helpful.  Worth making it more explicit?

Comment (richardesigns) [reply docs/spa/index.md:25]: Fair enough :-)

File: docs/spa/index.md (hunk 11 above)
Comment (richardesigns) [reply docs/spa/index.md:61]: Why not include that note?

File: docs/spa/index.md (hunk 17 above)
Comment (richardesigns) [reply docs/spa/index.md:143]: Ah right ok. This wasn't clear is my feedback but now I know it's an appendix then that's quite helpful. Worth making it more explicit?

Comment (CrowdHailer) [reply docs/spa/index.md:143]: How, would you do that?

**Pull request merged**